File: me/ch7/playlist.scala
// 1. Playlist has a name, a kind, and a list of songs.
// 2. There are three kinds of playlists: curated by a user, based on a
// particular artist, and based on a specific set of genres.
// 3. A song has an artist and a name.
// 4. A user has a name.
// 5. An artist has a name.
// 6. There are only three music genres: use your three favorite genres.

import scala.io.Codec.string2codec

object model {
  enum MusicGenre {
    case ROCK 
    case HIP_HOP
    case POP
  }

  enum Kind {
    case CuratedByUser(user: User)
    case BasedOnArtist(artist: Artist)
    case BasedOnGenres(genres: Set[MusicGenre])
  }

  opaque type Artist = String
  object Artist {
    def apply(value: String): Artist = value
    extension(a: Artist) def name: String = a
  }

  opaque type User = String
  object User {
    def apply(value: String): User = value
    extension(a: User) def name: String = a
  }

  case class Song(name: String, artist: Artist)

  case class Playlist(name: String, kind: Kind, songs: List[Song])

  def gatherSongs(playlists: List[Playlist], artist: Artist, genre: MusicGenre): List[Song] =
    playlists.foldLeft(List.empty[Song])((songs, playlist) =>
      val matchingSongs = playlist.kind match {
        case Kind.CuratedByUser(user) => playlist.songs.filter(_.artist == artist)
        case Kind.BasedOnArtist(playlistArtist) => if (playlistArtist == artist) playlist.songs
          else List.empty
        case Kind.BasedOnGenres(genres) => if (genres.contains(genre)) playlist.songs
          else List.empty
      }
      songs.appendedAll(matchingSongs)
  )


  // def gatherSongs(playlists: List[Playlist], artist: Artist, genre: MusicGenre): List[Song] = {
  //   playlists.flatMap(playlist => {
  //     playlist.kind match {
  //       case Kind.CuratedByUser(user) => playlist.songs.filter(_.artist == artist)
  //       case Kind.BasedOnArtist(_artist) => if (_artist == artist) playlist.songs else List.empty
  //       case Kind.BasedOnGenres(_genres) => if (_genres.contains(genre)) playlist.songs else List.empty
  //     }
  //   })
  // }
}

File: me/ch7/playlist.ts
import { iso, type Newtype } from "newtype-ts";
import { match } from "ts-pattern";

enum MusicGenre {
  HipHop,
  Pop,
  Rock,
}

interface User extends Newtype<{ readonly User: unique symbol }, string> {}
const isoUser = iso<User>();

interface Artist extends Newtype<{ readonly Artist: unique symbol }, string> {}
const isoArtist = iso<Artist>();

type PlaylistKind =
  | { type: "basedOnUser"; user: User }
  | { type: "basedOnArtist"; artist: Artist }
  | { type: "basedOnGenre"; genres: Set<MusicGenre> };

interface Song {
  artist: Artist;
  name: string;
}

interface Playlist {
  name: string;
  kind: PlaylistKind;
  songs: Song[];
}

const gatherSongs = (
  playlists: Playlist[],
  artist: Artist,
  genre: MusicGenre
): Song[] => {
  return playlists.flatMap((playlist) =>
    match(playlist.kind)
      .with({ type: "basedOnUser" }, () =>
        playlist.songs.filter((song) => song.artist === artist)
      )
      .with({ type: "basedOnArtist" }, ({ artist: _artist }) =>
        _artist == artist ? playlist.songs : []
      )
      .with({ type: "basedOnGenre" }, ({ genres: _genres }) =>
        _genres.has(genre) ? playlist.songs : []
      )
      .exhaustive()
  );
};

// const gatherSongsV2 = (
//   playlists: Playlist[],
//   artist: Artist,
//   genre: MusicGenre
// ): Song[] => {
//   return playlists.reduce(
//     (acc: Song[], playlist) =>
//       acc.concat(extractSongBaseOnKind(playlist, artist, genre)),
//     []
//   );
// };

// create three playlists base on different kinds
const playlists: Playlist[] = [
  {
    name: "HipHop Weekly",
    kind: { type: "basedOnGenre", genres: new Set([MusicGenre.HipHop]) },
    songs: [
      { artist: isoArtist.wrap("Drake"), name: "Views" },
      { artist: isoArtist.wrap("Bruno Mars"), name: "No Woman, No Cry" },
    ],
  },
  {
    name: "Rock Weekly",
    kind: { type: "basedOnArtist", artist: isoArtist.wrap("The Weeknd") },
    songs: [
      { artist: isoArtist.wrap("The Weeknd"), name: "After Hours" },
      { artist: isoArtist.wrap("Post Malone"), name: "Closer" },
    ],
  },
  {
    name: "Pop Weekly",
    kind: {
      type: "basedOnUser",
      user: isoUser.wrap("John Legend"),
    },
    songs: [
      { artist: isoArtist.wrap("Maroon 5"), name: "Sugar" },
      { artist: isoArtist.wrap("Avicii"), name: "Wake Me Up" },
    ],
  },
];

console.log(
  gatherSongs(playlists, isoArtist.wrap("The Weeknd"), MusicGenre.Rock)
);
